core/src/com/projectkorra/projectkorra/util/LightManager.java

```diff
@@ -0,0 +1,443 @@
+
+            BlockData waterloggedLightData = lightData.clone();
+            ((Waterlogged) waterloggedLightData).setWaterlogged(true);
+            waterloggedLightDataMap.put(level, waterloggedLightData);
+        }
+    }
+
+    /**
+     * Starts the light reverter task by scheduling it to run at a fixed rate of 50 milliseconds, or 1 tick.
+     * If the scheduler is already shut down, a new ScheduledThreadPoolExecutor is created with the number of available processors.
+     */
+    private void startLightReverter() {
+        if (scheduler.isShutdown()) {
+            scheduler = new ScheduledThreadPoolExecutor(Runtime.getRuntime().availableProcessors());
+        }
+
+        scheduler.scheduleAtFixedRate(this::revertExpiredLights, 0, 50, TimeUnit.MILLISECONDS);
+        scheduler.scheduleAtFixedRate(this::processBlockChanges, 0, 50, TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Fades out a light by decrementing its brightness by 1. The light will stop fading when its brightness
+     * reaches 0. The fade process is scheduled to run every 50 milliseconds, or 1 tick.
+     *
+     * @param lightData the LightData object containing the light's brightness, location, UUID, and ephemeral flag
+     */
+    private void fadeLight(LightData lightData) {
+        int brightness = lightData.brightness;
+
+        class TaskHolder {
+            ScheduledFuture<?> future;
+        }
+        TaskHolder taskHolder = new TaskHolder();
+
+        Runnable task = new Runnable() {
+            private int currentBrightness = brightness;
+
+            @Override
+            public void run() {
+                currentBrightness--;
+                if (currentBrightness > 0) {
+                    sendLightChange(lightData.location, currentBrightness, lightData.observers);
+                } else {
+                    revertLight(lightData);
+                    taskHolder.future.cancel(false);
+                }
+            }
+        };
+
+        taskHolder.future = scheduler.scheduleAtFixedRate(task, 0, 50, TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Sends a block change to the specified location. Brightness of 0 indicates that the light should be reverted.
+     *
+     * @param location   the location where the light change is to be sent
+     * @param brightness the brightness level of the light
+     * @param observers  the list of players who can see the light
+     */
+    private void sendLightChange(Location location, int brightness, Collection<? extends Player> observers) {
+        BlockData lightData = brightness > 0 ? getLightData(location, brightness) : getCurrentBlockData(location);
+        World targetWorld = location.getWorld();
+        double maxDistanceSquared = Math.pow(Bukkit.getServer().getViewDistance() * 16, 2);
+
+        observers.stream()
+                .filter(player -> player != null && player.isOnline() && !player.isDead() && player.getWorld().equals(targetWorld))
```

Comment:
I'm not sure if Spigot changed this, but wouldn't you have to compare `World` names rather than the object itself? I.e.: `player.getWorld.getName().equals(targetWorld.getName())`.

Reply:
That's a fair point. I would assume checking in memory is working ok, or has been, and based on the overall logic I don't foresee any issues with worlds state/unloading affecting it either. But if we want to be safe and check for equality in some way other than memory references, I would say compare the uuid of each world to be sure?

Reply:
I'm not sure you can compare UUIDs. I don't believe `World`s contain any.

> I would assume checking in memory is working ok, or has been

Yeah, probably. I think you're fine here if there's been no problems.